src/adapters/uniswap-v3/products/pool/uniswapV3PoolAdapter.ts

```diff
@@ -0,0 +1,348 @@
+        positionsManagerContract.positions(tokenId, { blockTag: blockNumber }),
+      ),
+    )
+
+    const nonZeroPositions = filterMap(positionMetadata, (position, index) => {
+      if (position.liquidity == 0n) return undefined
+
+      return {
+        liquidity: position.liquidity,
+        token0: position.token0,
+        token1: position.token1,
+        fee: position.fee,
+        tokenId: tokenIds[index],
+      }
+    })
+
+    if (nonZeroPositions.length == 0) return []
+
+    const tokenBalancesPromises = nonZeroPositions.map(
+      ({ liquidity, tokenId }) =>
+        positionsManagerContract.decreaseLiquidity.staticCallResult(
+          {
+            tokenId: tokenId as bigint,
+            liquidity,
+            amount0Min: 0n,
+            amount1Min: 0n,
+            deadline,
+          },
+          { from: userAddress, blockTag: blockNumber },
+        ) as unknown as Promise<
+          [bigint, bigint] & { amount0: bigint; amount1: bigint }
+        >, // IMO typechain has an bug here, incorrect type being returned
+    )
+
+    const claimableFeesPromises = nonZeroPositions.map(
+      ({ tokenId }) =>
+        positionsManagerContract.collect.staticCallResult(
+          {
+            tokenId: tokenId as bigint,
+            recipient: userAddress,
+            amount0Max: maxUint128,
+            amount1Max: maxUint128,
+          },
+          { from: userAddress, blockTag: blockNumber },
+        ) as unknown as Promise<
+          [bigint, bigint] & { amount0: bigint; amount1: bigint }
+        >, // IMO typechain has an bug here, incorrect type being returned
+    )
+
+    const erc20MetadataPromises = nonZeroPositions.map(({ token0, token1 }) =>
+      Promise.all([
+        getTokenMetadata(token0, this.chainId),
+        getTokenMetadata(token1, this.chainId),
+      ]),
+    )
+
+    const [tokenBalances, claimableFees, erc20Metadata] = await Promise.all([
+      Promise.all(tokenBalancesPromises),
+      Promise.all(claimableFeesPromises),
+      Promise.all(erc20MetadataPromises),
+    ])
+
+    return nonZeroPositions.map((pos, index) => {
+      const token0RawBalance = tokenBalances[index]!.amount0
+
+      const token1RawBalance = tokenBalances[index]!.amount1
+
+      const token0FeeRawBalance = claimableFees[index]!.amount0
+
+      const token1FeeRawBalance = claimableFees[index]!.amount1
+
+      const token0Index = 0
+      const token1Index = 1
+      const FEE_DECIMALS = 4
+      const nftName = `${erc20Metadata[index]![token0Index].symbol} / ${
+        erc20Metadata[index]![token1Index].symbol
+      } - ${formatUnits(pos.fee, FEE_DECIMALS)}%`
+      return {
+        address: positionManagerCommonAddress,
+        name: nftName,
+        symbol: nftName,
+        decimals: 18,
+        balanceRaw: pos.liquidity,
+        type: TokenType.Protocol,
+        tokens: [
+          this.createUnderlyingToken(
+            pos.token0,
+            erc20Metadata[index]?.[token0Index],
```

Comment:
If you know it cannot be undefined, then we should be using `erc20Metadata[index]!`